test/e2e/fqdn_dns_cache_test.go

```diff
@@ -0,0 +1,350 @@
+*/
+
+const (
+	testFullyQualifiedDomainName = "fqdn-test-pod.lfx.test"
+
+	customDnsServiceName = "custom-dns-service"
+	customDnsConfigName  = "custom-dns-config"
+	customDnsPort        = 53
+	customDnsTargetPort  = 53
+
+	customDnsImage         = "coredns/coredns:1.11.3"
+	customDnsPodName       = "custom-dns-server"
+	customDnsContainerName = "coredns"
+	customDnsLabelKey      = "app"
+	customDnsLabelValue    = "custom-dns"
+	customDnsVolume        = "config-volume"
+
+	fqdnPolicyName            = "test-anp-fqdn"
+	fqdnPodSelectorLabelKey   = "app"
+	fqdnPodSelectorLabelValue = "fqdn-cache-test"
+	toolBoxPodName            = "toolbox"
+
+	agnHostPort          = 80
+	agnHostLabelKey      = "app"
+	agnHostLabelValue    = "agnhost"
+	agnHostPodNamePreFix = "agnhost-"
+
+	eventualWaitTime = 20 * time.Second
+)
+
+func TestFQDNPolicyWithCachedDNS(t *testing.T) {
+	skipIfAntreaPolicyDisabled(t)
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+
+	customDnsService, err := data.CreateServiceWithAnnotations(customDnsServiceName, data.testNamespace, customDnsPort, customDnsTargetPort, corev1.ProtocolUDP,
+		map[string]string{customDnsLabelKey: customDnsLabelValue}, false, false, corev1.ServiceTypeClusterIP, ptr.To[corev1.IPFamily](corev1.IPv4Protocol), map[string]string{})
+	require.NoError(t, err, "Error when creating custom DNS service: %v", err)
+
+	setDnsServerAddressInAntrea(t, data, customDnsService.Spec.ClusterIP)
+
+	podCount := 2
+	agnHostPodIps := make([]*PodIPs, podCount)
+	for i := 0; i < podCount; i++ {
+		agnHostPodIps[i] = createHttpAgnhostPod(t, data, randName(agnHostPodNamePreFix), map[string]string{agnHostLabelKey: agnHostLabelValue})
+	}
+
+	ipv4, ipv6 := extractIPs(agnHostPodIps[0])
+
+	dnsConfigData := createDnsConfig(t, ipv4, ipv6)
+	customDnsConfigMapObject, err := data.CreateConfigMap(data.testNamespace, customDnsConfigName, dnsConfigData, nil, false)
+	require.NoError(t, err, "failed to create custom dns ConfigMap: %v", err)
+
+	createDnsPod(t, data)
+
+	k8sUtils, err = NewKubernetesUtils(data)
+	require.NoError(t, err, "error getting k8s utils %+v", err)
+
+	_ = buildFqdnPolicy(t, data)
+
+	createToolBoxPod(t, data, customDnsService.Spec.ClusterIP)
+
+	curlFqdn := func(podName, containerName, fqdn string) error {
+		cmd := []string{"curl", fqdn}
+		stdout, stderr, err := data.RunCommandFromPod(data.testNamespace, podName, containerName, cmd)
+		if err != nil {
+			return fmt.Errorf("error when running command '%s' on Pod '%s': %v, stdout: <%v>, stderr: <%v>",
+				strings.Join(cmd, " "), podName, err, stdout, stderr)
+		}
+		return nil
+	}
+
+	assert.Eventually(t, func() bool {
+		t.Logf("trying to curl the fqdn %v", testFullyQualifiedDomainName)
+		err = curlFqdn(toolBoxPodName, toolboxContainerName, testFullyQualifiedDomainName)
+		return assert.NoError(t, err)
+	}, 2*time.Second, 100*time.Millisecond)
```

Comment:
if you want to use `assert.NoError`, then you should `assert.EventuallyWithT`, and not `assert.Eventually`:

```go
assert.EventuallyWithT(t, func(t *assert.CollectT) {
        t.Logf("trying to curl the fqdn %v", testFullyQualifiedDomainName)
        assert.NoError(t, curlFqdn(toolBoxPodName, toolboxContainerName, testFullyQualifiedDomainName))
}, 2*time.Second, 100*time.Millisecond)
```

Refer to the testify docs for more details